Add support for TI CC3235sf

vanti opened the pull request:
This patch set adds support for the TI CC3235SF SoC and the CC3235SF LaunchPad (cc3235sf_launchxl).

Note that in order to flash and run code on the board, a file named 'ti_cc3220sf_vr.cfg' from TI OpenOCD is necessary. This file is still currently under review. The plan is to make this file publicly available in TI OpenOCD as soon as it is approved.

vanti committed:
samples: net: Adding board config files for CC3235SF

Config files are added to support cc3235sf_launchxl as a new board.

Signed-off-by: Vincent Wan <vincent.wan@linaro.org>

vanti committed:
boards: cc3235sf_launchxl: add support for TI CC3235SF LaunchPad

This commit adds the TI CC3235SF LaunchPad as a new board in Zephyr.

Signed-off-by: Vincent Wan <vincent.wan@linaro.org>

vanti committed:
soc: ti_simplelink: add CC3235SF to the CC32xx series

This commit adds the TI CC3235SF SoC to the CC32xx series.

Signed-off-by: Vincent Wan <vincent.wan@linaro.org>

vanti committed:
dts: arm: add device tree file for TI CC3235SF

This dtsi file adds definitions for memory regions on the SoC.

Signed-off-by: Vincent Wan <vincent.wan@linaro.org>

vanti committed:
ext: simplelink: Update comment for HAS_CC3220SDK

This Kconfig option applies to the TI CC32xx SDK in general, given it is
the same as the CC3220 SDK.

Signed-off-by: Vincent Wan <vincent.wan@linaro.org>

vanti committed:
linker: place debug header section for CC3235SF

The CC3235SF requires a debug header to be placed at the beginning of
internal flash, so that the bootloader does not automatically overwrite
the flash memory with the program saved in external flash.

Signed-off-by: Vincent Wan <vincent.wan@linaro.org>

GAnthony commented on boards/arm/cc3235sf_launchxl/support/openocd.cfg:
```diff
@@ -0,0 +1,10 @@
+source [find interface/xds110.cfg]
+transport select swd
+source [find target/ti_cc3220sf_vr.cfg]
+if { ![using_hla] } {
```
Was this bit taken from previous boards/arm/cc3220sf_launchxl/support/openocd.cfg?  If so, it was a workaround for the  cc3220sf reset issues, and may not apply for the cc3235sf_launchxl (unless that board has the same reset/debug issue). 
Could you check against the cfg file from the TI OpenOCD release for cc3235sf, and probably use that one instead?

vanti replied:
@GAnthony TI OpenOCD does not have any cfg file specific to the CC3235SF. So I have been using the same files as CC3220SF. If I revert to using openocd/tcl/board/ti_cc3220sf_launchpad.cfg, without your workaround, I see the following failure when running 'make flash':

> Using runner: openocd
> Open On-Chip Debugger 0.10.0
> Licensed under GNU GPL v2
> For bug reports, read
> 	http://openocd.org/doc/doxygen/bugs.html
> adapter speed: 2500 kHz
> srst_only separate srst_gates_jtag srst_open_drain connect_deassert_srst
> adapter_nsrst_delay: 1100
> Info : XDS110: connected
> Info : XDS110: firmware version = 2.3.0.17
> Info : XDS110: hardware version = 0x0027
> Info : XDS110: connected to target via SWD
> Info : XDS110: SWCLK set to 2500 kHz
> Info : clock speed 2500 kHz
> Info : SWD DPIDR 0x2ba01477
> Info : cc32xx.cpu: hardware has 6 breakpoints, 4 watchpoints
>     TargetName         Type       Endian TapName            State       
> --  ------------------ ---------- ------ ------------------ ------------
>  0* cc32xx.cpu         cortex_m   little cc32xx.dap         running
> Error: DP initialisation failed
> in procedure 'reset' 
> in procedure 'ocd_bouncer'
> 
> 
> Info : XDS110: disconnected

GAnthony replied:
OK, thanks for checking.

GAnthony commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -136,6 +136,13 @@ SECTIONS
 	KEEP(*(".dbghdr.*"))
 #endif
 
+#ifdef CONFIG_CC3235SF_DEBUG
```
Maybe combine with above, like:

```
# if defined(CONFIG_CC3220SF_DEBUG) || defined(CONFIG_CC3235SF_DEBUG)
[...]
#endif
```

dbkinder commented on boards/arm/cc3235sf_launchxl/doc/index.rst:
```diff
@@ -0,0 +1,305 @@
+********
+The SimpleLink Wi-Fi CC3235SF LaunchPad development kit (CC3235SF-LAUNCHXL)
+highlights CC3235SF, a single-chip wireless microcontroller (MCU) with
+1MB internal flash, 4MB external serial flash, 256KB of RAM and enhanced
```
```suggestion
1MB internal flash, 4MB external serial flash, 256KB of RAM, and enhanced
```

dbkinder commented on boards/arm/cc3235sf_launchxl/doc/index.rst:
```diff
@@ -0,0 +1,305 @@
+instructions on installation of tools, and how to flash the board using
+UniFlash.
+
+Note that zephyr.bin produced by the Zephyr SDK may not load via
```
```suggestion
Note that ``zephyr.bin`` produced by the Zephyr SDK may not load via
```

dbkinder commented on boards/arm/cc3235sf_launchxl/doc/index.rst:
```diff
@@ -0,0 +1,305 @@
+UniFlash.
+
+Note that zephyr.bin produced by the Zephyr SDK may not load via
+UniFlash tool.  If encountering difficulties, use the zephyr.elf
```
```suggestion
UniFlash tool.  If encountering difficulties, use the ``zephyr.elf``
```

dbkinder commented on boards/arm/cc3235sf_launchxl/doc/index.rst:
```diff
@@ -0,0 +1,305 @@
+
+   Download and install the latest `XDS-110 emulation package`_.
+
+   Follow the directions here to update the firmware:
```
This would read better as:
```
   Follow these `xds110 firmware update directions
   <http://software-dl.ti.com/ccs/esd/documents/xdsdebugprobes/emu_xds110.html#updating-the-xds110-firmware>`_
```

dbkinder commented on boards/arm/cc3235sf_launchxl/doc/index.rst:
```diff
@@ -0,0 +1,305 @@
+   http://software-dl.ti.com/ccs/esd/documents/xdsdebugprobes/emu_xds110.html#updating-the-xds110-firmware
+
+   Note that the emulation package install may place the xdsdfu utility
+   in <install_dir>/ccs_base/common/uscif/xds110/.
```
```suggestion
   in ``<install_dir>/ccs_base/common/uscif/xds110/``.
```

dbkinder commented on boards/arm/cc3235sf_launchxl/doc/index.rst:
```diff
@@ -0,0 +1,305 @@
+   Follow the instructions in the Release Notes in that repository to build
+   and install.
+
+   Since the default TI OpenOCD installation is /usr/local/bin/,
```
```suggestion
   Since the default TI OpenOCD installation is ``/usr/local/bin/``,
```

dbkinder commented on boards/arm/cc3235sf_launchxl/doc/index.rst:
```diff
@@ -0,0 +1,305 @@
+   and install.
+
+   Since the default TI OpenOCD installation is /usr/local/bin/,
+   and /usr/local/share/, you may want to backup any current openocd
```
```suggestion
   and ``/usr/local/share/``, you may want to backup any current openocd
```

dbkinder commented on boards/arm/cc3235sf_launchxl/doc/index.rst:
```diff
@@ -0,0 +1,305 @@
+   :goals: debug
+
+
+WiFi Support
```
Wi-Fi is the correct spelling of the trademarked name. Fix all text references (there are few more WiFi uses below).

vanti commented:
Rebased, and updated based on comments from @GAnthony and @dbkinder. Thanks.

I have tweaked the Kconfig files of the samples to enable "CONFIG_CC3235SF_DEBUG". This helps make their behavior more consistent, by preventing the internal flash from being overwritten by contents from external flash. I have also removed 'CONFIG_NET_SOCKETS_SOCKOPT_TLS=y' for http_get so that TLS is only enabled when the overlay for TLS is included in the build.

vanti commented:
@dbkinder  Let me know if more changes are needed for approval.

dbkinder reviewed (approved):
doc changes LGTM, thanks!

vanti commented:
Need one more approval from reviewers

GAnthony reviewed (approved):
Need to check the licensing failure on shippable, otherwise code looks good.

vanti commented:
> Need to check the licensing failure on shippable, otherwise code looks good.

Thanks @GAnthony. The file is exactly the same as the one for CC3220SF in boards/arm/cc3220sf_launchxl/pinmux.c, hence should carry the same license.

galak commented on boards/arm/cc3235sf_launchxl/board.cmake:
```diff
@@ -0,0 +1,8 @@
+# Use the TI OpenOCD (by default in /usr/local/openocd)
+# See the Zephyr project CC3235SF_LAUNCHXL documentation on
+# flashing prerequisites.
+set(OPENOCD "/usr/local/bin/openocd" CACHE FILEPATH "" FORCE)
```
We should look at using the Zephyr SDK for this.

vanti replied:
@galak I tried OpenOCD in the new Zephyr SDK 0.10.1 today, and found that the wifi driver fails to start on the CC3235SF if flashed using it. I don't see this problem when using TI OpenOCD.

galak replied:
If you reset the board after flashing with the Zephyr SDK do things work?

galak replied:
Can we try SDK 0.10.2 and see how things work.  Even with the reset issue, I'd prefer we use stock zephyr openocd here.

galak commented on boards/arm/cc3235sf_launchxl/pinmux.c:
```diff
@@ -0,0 +1,167 @@
+/*
+ * pinmux.c
+ *
+ * configure the device pins for different peripheral signals
```
is there a reason this isn't apache licensed?

vanti replied:
@galak I think it should be fine, but given it's another copy of a file that already exists, I didn't change the license. If you think it makes sense to change it to Apache I can modify it.

galak reviewed (changes requested):
Some minor changes related to OpenOCD and license of one file.

vanti commented:
Rebased and modified licensing in pinmux.c.

jukkar commented:
@galak can you recheck the fixes, the issues seems to be addressed.

pfalcon commented:
> @galak can you recheck the fixes, the issues seems to be addressed.

Ping @galak 

@vanti: Can you in the meantime rebase on the latest master?

pfalcon commented on soc/arm/ti_simplelink/cc32xx/Kconfig.soc:
```diff
@@ -21,3 +25,11 @@ config CC3220SF_DEBUG
+
+if SOC_CC3235SF
+
+config CC3235SF_DEBUG
```
Is this the right option name? Wouldn't a more specific, like "CC3235SF_DEBUG_IMG_HDR" would be better?

pfalcon reviewed (approved):
Looks good. I assume OpenOCD matters can be elaborated later.

ioannisg commented:
@galak pls, take a look (CI seems green now)

vanti committed:
boards: cc3235sf_launchxl: use OpenOCD from Zephyr SDK

We should use OpenOCD from the latest Zephyr SDK as opposed to using TI
OpenOCD.

There is a known issue with the current release (0.10.2) of the SDK in
how it resets the network processor, but we can deal with it
subsequently.

Signed-off-by: Vincent Wan <vincent.wan@linaro.org>

vanti reviewed (commented):
@galak @pfalcon I have rebased and updated the cmake files to use openocd from the SDK, and added some documentation that the reset button needs to be manually hit after flashing code that makes use of the network processor when using SDK 0.10.2.

vanti commented:
Since this PR is marked for 2.1.0, I am going to submit 
commit 99ea5f90ad3cda2acff36e6df651d544cb0a10ae to fix the log_strdup in a separate PR, so that it can be addressed for CC3220SF in 2.0.0.

vanti commented:
> Since this PR is marked for 2.1.0, I am going to submit
> commit [99ea5f9](https://github.com/zephyrproject-rtos/zephyr/commit/99ea5f90ad3cda2acff36e6df651d544cb0a10ae) to fix the log_strdup in a separate PR, so that it can be addressed for CC3220SF in 2.0.0.

Put up PR #18562.

jukkar commented:
@vanti could you rebase against latest master so we can merge this?

galak commented:
@vanti can you look at the build failure.

vanti committed:
samples: net: sockets: disable http_get in test build

Build errors were introduced by the merge of #18736. Until PR #18780 is
approved to allow the SimpleLink libraries to build without
CONFIG_POSIX_API, this patch excludes cc3235sf_launchxl from the test
build.

Signed-off-by: Vincent Wan <vincent.wan@linaro.org>

vanti commented:
> @vanti can you look at the build failure.

@galak This is the same failure that #18780 is trying to address on the cc3220sf_launchxl. Until that PR is approved, I am disabling the build for the http_get sample in sample.yaml, similar to what was done for the other board.

@jukkar Rebased on master, thanks.
